pkg/drift/driftignore.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package drift
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/abcxyz/pkg/logging"
+
+	"github.com/abcxyz/guardian/pkg/drift/assets"
+	"github.com/abcxyz/guardian/pkg/drift/iam"
+)
+
+type ignoredAssets struct {
+	iamAssets map[string]struct{}
+	projects  map[string]struct{}
+	folders   map[string]struct{}
+}
+
+type hierarchyNode struct {
+	*assets.HierarchyNode
+	projectIDs []string
+	folderIDs  []string
+}
+
+// ignoredProjectPattern is a Regex pattern used to identify projects that should be ignored.
+var ignoredProjectPattern = regexp.MustCompile(`^\/organizations\/(?:\d*)\/projects\/(\d*)$`)
+
+// ignoredFolderPattern is a  Regex pattern used to identify folders that should be ignored.
+var ignoredFolderPattern = regexp.MustCompile(`^\/organizations\/(?:\d*)\/folders\/(\d*)$`)
+
+func filterIgnored(values map[string]*iam.AssetIAM, ignored *ignoredAssets) map[string]*iam.AssetIAM {
+	filtered := make(map[string]*iam.AssetIAM)
+	for k, a := range values {
+		if a.ResourceType == assets.Project {
+			if _, ok := ignored.projects[a.ResourceID]; !ok {
+				filtered[k] = a
+			}
+		} else if a.ResourceType == assets.Folder {
+			if _, ok := ignored.folders[a.ResourceID]; !ok {
+				filtered[k] = a
+			}
+		} else { // Handle default case so we do not accidentally drop values.
+			filtered[k] = a
+		}
+	}
+
+	return filtered
+}
+
+func expandGraph(ignored *ignoredAssets, hierarchyGraph map[string]*hierarchyNode) (*ignoredAssets, error) {
```

Comment:
I think we really need some unit coverage for this function and `foldersBeneath`.

Reply:
Oh yeah we do

Reply:
Moved to the assets package and added some tests.